IdentityCore/src/oauth2/MSIDTokenResponseHandler.m

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDTokenResponseHandler.h"
+#import "MSIDAADTokenResponse.h"
+#import "MSIDRequestContext.h"
+#import "MSIDAADV1TokenResponse.h"
+#import "MSIDAADV2TokenResponse.h"
+#import "MSIDError.h"
+#import "MSIDClientInfo.h"
+#import "MSIDAADV2RequestParameters.h"
+#import "NSOrderedSet+MSIDExtensions.h"
+
+@implementation MSIDTokenResponseHandler
+
++ (BOOL)processResponse:(MSIDAADTokenResponse *)response
+           refreshToken:(NSString *)refreshToken
+          requestParams:(MSIDRequestParameters *)parameters
+                context:(id<MSIDRequestContext>)context
+                  error:(NSError * __autoreleasing *)error
+{
+    if (!response)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorInternal, @"processTokenResponse called without a response dictionary", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    [self.class checkCorrelationId:response requestCorrelationId:context.correlationId];
+    
+    if (response.error)
+    {
+        if (error)
+        {
+            MSIDErrorCode errorCode = refreshToken ? MSIDErrorServerRefreshTokenRejected : MSIDErrorServerOauth;
+            if ([response isKindOfClass:MSIDAADV2TokenResponse.class])
+            {
+                errorCode = [self.class getErrorCodeForAADV2:response.error];
```

Comment:
this looks like something that can be moved to the base token response class and override by its subclasses to avoid this check

Reply:
done